modules/gdscript/gdscript_format.cpp

```diff
@@ -0,0 +1,1958 @@
+/* Copyright (c) 2007-2022 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "gdscript_format.h"
+#ifdef TOOLS_ENABLED
+#include "editor/editor_settings.h"
+#endif
+#include "gdscript_parser.h"
+
+GDScriptFormat::GDScriptFormat() :
+		line_length_maximum(100),
+		tab_size(4),
+		tab_type(0),
+		lines_between_functions(2),
+		indent_in_multiline_block(2) {
+#ifdef TOOLS_ENABLED
+	if (EditorSettings::get_singleton()) {
+		line_length_maximum = EditorSettings::get_singleton()->get_setting("text_editor/appearance/guidelines/line_length_guideline_hard_column");
+		lines_between_functions = EditorSettings::get_singleton()->get_setting("text_editor/behavior/formatter/lines_between_functions");
+		indent_in_multiline_block = EditorSettings::get_singleton()->get_setting("text_editor/behavior/formatter/indent_in_multiline_block");
+		tab_size = EditorSettings::get_singleton()->get_setting("text_editor/behavior/indent/size");
+		tab_type = EditorSettings::get_singleton()->get_setting("text_editor/behavior/indent/type");
+	}
+#endif // TOOLS_ENABLED
+}
+
+String GDScriptFormat::format(const String &p_code) {
+	GDScriptParser parser;
+	parser.parse(p_code, "", false);
+
+	if (!parser.get_errors().is_empty()) {
+		return p_code;
+	}
+
+	find_custom_newlines(p_code);
+
+	GDP::ClassNode *root = parser.get_tree();
+
+	String output = "";
```

Comment:
With all this string manipulation, would it be worth looking at `StringBuilder`?